### PR TITLE
Drop PHP 7.4 support, upgrade to vimeo/psalm v5

### PR DESCRIPTION
**Pull request by internalsystemerror**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | yes

### Description

This PR:
- Adds support for PHP 8.2
- Drops support for PHP 7.4
- Updates `vimeo/psalm` to v5
- Solves various psalm errors (but also creates a couple of new ones added to the baseline sadly).

Closes:
- https://github.com/laminas/laminas-cache-storage-adapter-redis/pull/57
- https://github.com/laminas/laminas-cache-storage-adapter-redis/pull/58

### CODE REVIEW

**Comment by internalsystemerror** [comment]
There are various errors shown in CI here:
- `boesing/psalm-plugin-stringf` not supporting PHP 8.2 (solved by: https://github.com/boesing/psalm-plugin-stringf/pull/142)
- Too many open files (not sure what is causing this still)
- `RedisCluster::keys(): Can't send KEYS to 172.18.0.3:6380` (which I'm unable to test locally as I don't have redis cluster setup).

Hopefully this can at least be a place to start if anyone else wishes to pick this up before I can fix, but those last 2 are really stumping me.

**Comment by boesing** [comment]
I already spent some time in trying to fix CI regarding Open Files and KEYS error.

Will see that the open PRs get merged once I get CI running. I do prefer not merging this without CI passing for the unit tests.


Too many open files is the error I mentioned in slack. Its ridicculous and drives me crazy as all CI errors we have right now are hard to tackle...

**Comment by internalsystemerror** [comment]
The open files issue only occurs with the latest dependencies, if I could reproduce that locally (which I know you mentioned you can't either) then we might be able to figure out which version of which dep and conflict with it. Or even open an issue upstream.

**Comment by boesing** [comment]
@internalsystemerror I am currently merging the renovate PRs. Are you open to apply your changes to the renovate PR regarding psalm v5?

**Comment by internalsystemerror** [comment]
I prefer not to work directly on the renovate branches myself as it prevents renovate from updating them in the future so its a habit I've gotten into. But I can rebase this one using the commit from the renovate branch tomorrow.

File: src/RedisResourceManager.php [hunk 1]
```diff
@@ -233,9 +233,8 @@ protected function normalizeServer(&$server)
                 throw new Exception\InvalidArgumentException("Invalid server given");
             }
 
-            $host    = $server['host'];
-            $port    = isset($server['port']) ? (int) $server['port'] : $port;
-            $timeout = isset($server['timeout']) ? (int) $server['timeout'] : $timeout;
```
Comment (Ocramius) [inline src/RedisResourceManager.php:238]: This code block is not unused? 🤔

File: src/RedisResourceManager.php [hunk 2]
```diff
@@ -596,7 +595,7 @@ protected function normalizeLibOptionKey(&$key)
             throw new Exception\InvalidArgumentException("Unknown redis option '{$key}' ({$const})");
         }
         /** @var int $key */
-        $key = constant($const);
+        $key = (int) constant($const);
```
Comment (Ocramius) [inline src/RedisResourceManager.php]: IMO a cast is worse here: it obfuscates cases where non-numeric values are fetched

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by internalsystemerror** [comment]
I've rebased but now I'm getting a load of extra psalm errors. I think the options array needs typing and I'm kinda confused what's expected, e.g. `$server['timeout']` is used where `$server['port']` should be, things like that.

File: src/RedisResourceManager.php (hunk 2 above)
Comment (internalsystemerror) [reply src/RedisResourceManager.php]: All these constant values should be an integer I believe

File: src/RedisResourceManager.php (hunk 1 above)
Comment (internalsystemerror) [reply src/RedisResourceManager.php:238]: 🤷

File: src/RedisResourceManager.php (hunk 2 above)
Comment (Ocramius) [reply src/RedisResourceManager.php]: Then we should at least `assert()` instead :-)

File: src/RedisResourceManager.php (hunk 1 above)
Comment (Ocramius) [reply src/RedisResourceManager.php:238]: Why did you drop it then? :D

Comment (internalsystemerror) [reply src/RedisResourceManager.php:238]: It is actually unused as `$server['timeout']` is never available at this point. To get here `$server` must be a string that is then turned into array via `parse_url` containing no timeout key, or a hard `['host' => ...]` again without any timeout key.

Comment (Ocramius) [reply src/RedisResourceManager.php:238]: aha, I see: so `$server['timeout']` can never exist here 👍

File: src/RedisResourceManager.php (hunk 2 above)
Comment (internalsystemerror) [reply src/RedisResourceManager.php]: Yeah I'd like to do much much more to ensure these types, and I think we may need to just accept BC breaks and a new major to do so.

File: src/RedisResourceManager.php (hunk 1 above)
Comment (internalsystemerror) [reply src/RedisResourceManager.php:238]: Exactly, sorry I misunderstood your first comment here.

File: src/RedisResourceManager.php (hunk 2 above)
Comment (internalsystemerror) [reply src/RedisResourceManager.php]: e.g. `normalizePersistentId` method just casts the variable reference to a string and nothing else, when the param type is already set to be a string. So assuming that the types are correct, this method is redundant, but that may actually break usages which aren't currently checking for types.

Comment (Ocramius) [reply src/RedisResourceManager.php]: for now I'd really just recommend coding the assumption in form an assertion though

Comment (internalsystemerror) [reply src/RedisResourceManager.php]: Will do for this line, but isn't that already a BC Break in case any of these aren't an integer?

**Comment by internalsystemerror** [comment]
Yeah, so to be specific. The issue is because in a lot of places rather than throwing exceptions, values are being coerced via normalisation which means to fix most of these issues is a BC break as far as I can see. If using specific types the vast majority of this normalisation becomes redundant also.

**Review by Ocramius — APPROVED** [review]
Thanks @internalsystemerror!

**Pull request merged**